### ISSUE TITLE
Odd toplevel printing for lists with local opens

### ISSUE THREAD

**Issue by yallop** [ocaml/ocaml#12966]
I'm not sure whether this should be considered a bug, but it's certainly surprising behaviour.  In OCaml 4.07.1, printing lists in the toplevel works as I'd expect; in particular, it's unaffected by local opens:

```
        OCaml version 4.07.1

# [1];;
- : int list = [1]
# List.[1];;
- : int list = [1]
```

In OCaml 4.08.0 onwards, lists are printed oddly (with qualified prefix cons, but unqualified nil)  when constructed below a local open:

```
        OCaml version 4.08.0

# [1];;
- : int list = [1]
# List.[1];;
- : int List.t = List.(::) (1, [])
```

**Comment by Octachron** [comment]
The special printing for list in the toplevel is triggered by recognizing the exact type path for list, which is very brittle
and breaks when re-exporting the list type:
```ocaml
module List = struct
  type 'a t = 'a list = [] | (::) of 'a * 'a list
 end
 open List
 let x = [1;2]
 ```
 > val x : int List.t = List.(::) (1, [2])

At the same time, making sure that the toplevel value printer always pick the best valid syntactic form in the current environment is not completely trivial.

**Comment by gasche** [comment]
@Octachron tried to do something "good" for this check in #10386, but I don't remember why it went nowhere -- except maybe "it's complex and no one knowledgeable was around to review it, only @gasche".

One compromise I proposed to @Octachron today is to at least normalize the type path when we do the comparison "is this the list type?", namely `Path.same path Predef.path_list` in genprintval.ml. But then there are many places in the compiler that use this same `Path.same foo Predef.bar` pattern, and I don't understand why we would want to use a more powerful equivalence here but not elsewhere.

**Comment by samsa1** [comment]
Is this solved by #13053 or is it still open ?

**Comment by Octachron** [comment]
This specific issue is indeed fixed by your PR, thanks for reminding this to me !
